H levels pr

IanRay11 opened the pull request:
For context: I am in the process of splitting up the PR #294 into more digestible pieces for easier review. Unfortunately, I created another upstream and with that a new PR so we lost the initial round of suggestions. I addressed those comments in my most recent commit.

tomdjong commented:
@IanRay11 Were you done making changes now?

IanRay11 commented:
Ready for further review

tomdjong commented on source/UF/H-Levels.lagda:
```diff
@@ -163,33 +196,6 @@ The subuniverse of types of hlevel n is defined as follows.
-
-\end{code}
-
 From Univalence we can show that (ℍ n) is of level (n + 1), for all n : ℕ.
```
univalence (lowercase)

tomdjong commented on source/UF/H-Levels.lagda:
```diff
@@ -202,69 +208,32 @@ From Univalence we can show that (ℍ n) is of level (n + 1), for all n : ℕ.
+                  (hlevels-are-upper-closed zero X' X'-is-contr))
     C' : is-contr (X ≃ X')
     C' = pointed-props-are-singletons (singleton-≃ X-is-contr X'-is-contr) P
 ℍ-is-of-next-hlevel (succ n) 𝓤 ua (X , l) (X' , l') =
```
I think you can clean up `ℍ-is-of-next-hlevel` quite a bit by first proving the auxiliary lemma
```
equiv-preserves-hlevel : (n : ℕ) {X : 𝓤 ̇  } {Y : 𝓥 ̇  }
                       → X is-of-hlevel n
                       → Y is-of-hlevel n
                       → (X ≃ Y) is-of-hlevel n
equiv-preserves-hlevel zero = {!!}
equiv-preserves-hlevel (succ n) = {!!}
```
Then you can prove `ℍ-is-of-next-hlevel` without pattern matching on `n`. It also removes the some of the duplication in the correct argument.

tomdjong replied:
Note that the zero case follows immediately from `singleton-≃`.

IanRay11 replied:
Shouldn't the fact that an equivalence between singletons is itsself a singleton already exist elsewhere?

IanRay11 replied:
To clarify I am fine with inlining it in my lemma but it just seems to be something that one would want developed in the general theory.

tomdjong replied:
> Shouldn't the fact that an equivalence between singletons is itself a singleton already exist elsewhere?

Yes, it's `singleton-≃` as I mentioned.

IanRay11 replied:
That simply gives the center not the contraction.

tomdjong replied:
Ah, sorry. Then I'll add it tomorrow (I'm afk atm).

martinescardo replied:
I think this is somewhere in TypeTopology. I'll try to find it later.

tomdjong replied:
@IanRay11 It's now merged as `≃-is-singleton` in `UF.Singleton-Properties`.

IanRay11 replied:
Thank you @tomdjong!

IanRay11 commented on source/UF/H-Levels.lagda:
```diff
@@ -33,41 +34,78 @@ open import UF.Subsingletons-FunExt
+is-prop'-implies-is-prop : {X : 𝓤 ̇} → is-prop' X → is-prop X
+is-prop'-implies-is-prop X-is-prop' x x' = center (X-is-prop' x x')
+
+is-prop-equiv-is-prop' : {𝓤 : Universe} {X :  𝓤 ̇} → is-prop X ≃ is-prop' X
```
Ahh I don't need it listed but I do need to implicilty supply it.

tomdjong replied:
You didn't give the error message, but I'm assuming it's only because you have `{𝓤}` as an argument to `is-prop-equiv-is-prop'`, but you can simply it when you delete it from the type signature. The rest of that code doesn't mention `𝓤` at all.

tomdjong replied:
> Ahh I don't need it listed but I do need to implicilty supply it.

I don't think so. None of the code mentions `𝓤`, so why would you need it? Please fix the indentation to be a single space by the way.

tomdjong replied:
Oh, I see what you mean now. Yes, you need to implicitly supply it because `X` is also implicit. Sorry for the noise.

tomdjong commented:
@IanRay11 Just to check: you're planning on making more changes before merging/further review, right?

IanRay11 commented:
@tomdjong yes I've got one more thing to address

IanRay11 commented:
@tomdjong okay I've addressed all suggestions. You can do another round of review, but note that in the meantime I will be adding the embedding that I use in the second to last proof to the embeddings file and plan to remove it from this file -- so no need to comment on that.

tomdjong commented on source/UF/H-Levels.lagda:
```diff
@@ -163,108 +193,34 @@ The subuniverse of types of hlevel n is defined as follows.
- infix 0 ∣∣_∣∣_
- infix 0 ∣_∣_
-
+  I = ≃-sym (to-subtype-＝-≃ (λ - → hlevel-relation-is-prop n -))
```
Please swap the labels `I` and `II` and then put `II` in a `where`-clause of `I` so that its type is clear.

tomdjong commented:
@IanRay11 Thanks Ian! After the embedding thing gets merged and used here (and the above comment is addressed), I think this is ready to get merged. Thanks for your patience and changes by the way. I hope you'll agree that the current proofs are easier to read :slightly_smiling_face:

IanRay11 commented:
@tomdjong oh they look much better!

IanRay11 commented:
@tomdjong I updated the file to use the embedding result

martinescardo commented:
I've had a 15min-time look at this last version, and I am happy with it.

But I think you should have the privilege of merging it, as you did the hard work of reviewing it earlier and giving advice.
